extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisBackedBlockRDD.scala

```diff
@@ -0,0 +1,223 @@
+    } else {
+      getBlockFromKinesis()
+    }
+  }
+}
+
+
+/** An iterator that return the Kinesis data based on the given range of Sequence numbers */
+private[kinesis]
+class KinesisSequenceRangeIterator(
+    credentials: AWSCredentials,
+    endpointUrl: String,
+    regionId: String,
+    range: SequenceNumberRange
+  ) extends NextIterator[Array[Byte]] {
+
+  private val backoffTimeMillis = 1000
+  private val client = new AmazonKinesisClient(credentials)
+
+  private var toSeqNumberReceived = false
+  private var lastSeqNumber: String = null
+  private var internalIterator: Iterator[Record] = null
+
+  client.setEndpoint(endpointUrl, "kinesis", regionId)
+
+  override protected def getNext(): Array[Byte] = {
+    var nextBytes: Array[Byte] = null
+    if (toSeqNumberReceived) {
+      finished = true
+    } else {
+
+      if (internalIterator == null) {
+
+        // If the internal iterator has not been initialized,
+        // then fetch records from starting sequence number
+        getRecords(ShardIteratorType.AT_SEQUENCE_NUMBER, range.fromSeqNumber)
+      } else if (!internalIterator.hasNext) {
+
+        // If the internal iterator does not have any more records,
+        // then fetch more records after the last consumed sequence number
+        getRecords(ShardIteratorType.AFTER_SEQUENCE_NUMBER, lastSeqNumber)
+      }
+
+      if (!internalIterator.hasNext) {
+
+        // If the internal iterator still does not have any data, then throw exception
+        // and terminate this iterator
+        finished = true
+        throw new SparkException(s"Could not read until the specified end sequence number: $range")
+      } else {
+
+        // Get the record, and remember its sequence number
+        val nextRecord = internalIterator.next()
+        nextBytes = nextRecord.getData().array()
+        lastSeqNumber = nextRecord.getSequenceNumber()
+
+        // If the this record's sequence number matches the stopping sequence number, then make sure
+        // the iterator is marked finished next time getNext() is called
+        if (nextRecord.getSequenceNumber == range.toSeqNumber) {
+          toSeqNumberReceived = true
+        }
+      }
+
+    }
+    nextBytes
+  }
+
+  override protected def close(): Unit = { }
+
+  private def getRecords(iteratorType: ShardIteratorType, seqNum: String): Unit = {
+    val shardIterator = getKinesisIterator(range.streamName, range.shardId, iteratorType, seqNum)
+    var records: Seq[Record] = null
+    do {
+      try {
+        val getResult = getRecordsAndNextKinesisIterator(
+          range.streamName, range.shardId, shardIterator)
+        records = getResult._1
+      } catch {
+        case ptee: ProvisionedThroughputExceededException =>
+          Thread.sleep(backoffTimeMillis)
+      }
+    } while (records == null || records.length == 0)  // TODO: put a limit on the number of retries
+    if (records != null && records.nonEmpty) {
```

Comment:
Is this condition necessary? Looks it's always true when exiting the `while` loop.